File: InMotionGIT.Common/Extensions/ListExtensions.vb
Imports System.Runtime.CompilerServices

Namespace Extensions

    ''' <summary>
    ''' Extension methods for the list data type
    ''' </summary>
    Public Module ListExtensions

        ''' <summary>
        ''' Defines whether a list is empty or contains elements within it/ Define si una lista es empty or no contiene elementos dentro de ella
        ''' </summary>
        ''' <typeparam name="T"></typeparam>
        ''' <param name="source"></param>
        ''' <returns></returns>
        <Extension()>
        Public Function IsEmptyAndNotContainsItems(Of T)(ByVal source As List(Of T)) As Boolean
            Dim result As Boolean = False
            If IsNothing(source) Then
                Return True
            End If
            If Not source.Count <> 0 Then
                Return True
            End If
            Return result
        End Function

        ''' <summary>
        ''' Creates a string separated by the character indicated property values that difinio/Crea un string separado por el carácter que se indique de  los valores de la propiedad que se difinio
        ''' </summary>
        ''' <typeparam name="T"></typeparam>
        ''' <param name="source"></param>
        ''' <param name="separator"></param>
        ''' <param name="propertyName"></param>
        ''' <returns></returns>
        <Extension()>
        Public Function CreateConcatenation(Of T)(ByVal source As List(Of T), separator As String, propertyName As String) As String
            Dim result As String = String.Empty
            If Not IsNothing(source) AndAlso source.Count <> 0 Then
                Dim temporalObject As Object = source.FirstOrDefault()
                If Not IsNothing(temporalObject) Then
                    Dim isExistProperty As Boolean = ObjectExtensions.ExistsProperty(temporalObject, propertyName)
                    If isExistProperty Then
                        Dim listaVector As New List(Of String)
                        For Each ItemSource In source
                            listaVector.Add(ItemSource.GetType().GetProperty(propertyName).GetValue(ItemSource, Nothing))
                        Next
                        If Not listaVector.IsEmptyAndNotContainsItems Then
                            result = String.Join(separator, listaVector)
                        End If
                    End If
                End If
            End If
            Return result
        End Function

        <Extension()>
        Public Sub AddByDistinct(ByRef source As List(Of String), value As String)
            If source.IsEmpty Then
                source = New List(Of String)
            End If
            Dim foundItem = (From itemDb In source Where itemDb.Equals(value) Select itemDb).FirstOrDefault()
            If foundItem.IsEmpty Then
                source.Add(value)
            End If
        End Sub

        <Extension()>
        Public Sub AddByDistinct(ByRef source As List(Of Int32), value As Int32)
            If source.IsEmpty Then
                source = New List(Of Int32)
            End If
            Dim foundItem = (From itemDb In source Where itemDb = value Select itemDb).FirstOrDefault()
            If foundItem.IsEmpty Then
                source.Add(value)
            End If
        End Sub

        <Extension()>
        Public Sub AddByDistinct(ByRef source As List(Of Int64), value As Int64)
            If source.IsEmpty Then
                source = New List(Of Int64)
            End If
            Dim foundItem = (From itemDb In source Where itemDb = value Select itemDb).FirstOrDefault()
            If foundItem.IsEmpty Then
                source.Add(value)
            End If
        End Sub

    End Module

End Namespace

File: Common/InMotionGIT.Common/Extensions/ListExtensions.cs
using System.Collections.Generic;
using System.Linq;
using Microsoft.VisualBasic.CompilerServices;

namespace InMotionGIT.Common.Extensions
{

    /// <summary>
    /// Extension methods for the list data type
    /// </summary>
    public static class ListExtensions
    {

        /// <summary>
        /// Defines whether a list is empty or contains elements within it/ Define si una lista es empty or no contiene elementos dentro de ella
        /// </summary>
        /// <typeparam name="T"></typeparam>
        /// <param name="source"></param>
        /// <returns></returns>
        public static bool IsEmptyAndNotContainsItems<T>(this List<T> source)
        {
            bool result = false;
            if (source == null)
            {
                return true;
            }
            if (!(source.Count != 0))
            {
                return true;
            }
            return result;
        }

        /// <summary>
        /// Creates a string separated by the character indicated property values that difinio/Crea un string separado por el carácter que se indique de  los valores de la propiedad que se difinio
        /// </summary>
        /// <typeparam name="T"></typeparam>
        /// <param name="source"></param>
        /// <param name="separator"></param>
        /// <param name="propertyName"></param>
        /// <returns></returns>
        public static string CreateConcatenation<T>(this List<T> source, string separator, string propertyName)
        {
            string result = string.Empty;
            if (!(source == null) && source.Count != 0)
            {
                object temporalObject = source.FirstOrDefault();
                if (!(temporalObject == null))
                {
                    bool isExistProperty = temporalObject.ExistsProperty(propertyName);
                    if (isExistProperty)
                    {
                        var listaVector = new List<string>();
                        foreach (var ItemSource in source)
                            listaVector.Add(Conversions.ToString(ItemSource.GetType().GetProperty(propertyName).GetValue(ItemSource, null)));
                        if (!listaVector.IsEmptyAndNotContainsItems())
                        {
                            result = string.Join(separator, listaVector);
                        }
                    }
                }
            }
            return result;
        }

        public static void AddByDistinct(ref List<string> source, string value)
        {
            if (source.IsEmpty())
            {
                source = new List<string>();
            }
            string foundItem = (from itemDb in source
                                where itemDb.Equals(value)
                                select itemDb).FirstOrDefault();
            if (foundItem.IsEmpty())
            {
                source.Add(value);
            }
        }

        public static void AddByDistinct(ref List<int> source, int value)
        {
            if (source.IsEmpty())
            {
                source = new List<int>();
            }
            int foundItem = (from itemDb in source
                             where itemDb == value
                             select itemDb).FirstOrDefault();
            if (foundItem.IsEmpty())
            {
                source.Add(value);
            }
        }

        public static void AddByDistinct(ref List<long> source, long value)
        {
            if (source.IsEmpty())
            {
                source = new List<long>();
            }
            long foundItem = (from itemDb in source
                              where itemDb == value
                              select itemDb).FirstOrDefault();
            if (foundItem.IsEmpty())
            {
                source.Add(value);
            }
        }

    }

}